### PR TITLE
Ldap Suppport with Groups

### LINKED ISSUE #101: LDAP support
I found kafkahq could outperform kafka-manager more, if LDAP support is enabled as kafka-manager does

https://github.com/yahoo/kafka-manager#authenticating-a-user-with-ldap

Is the LDAP support already available via configuration and without any code change? as the micronaut document written here
https://micronaut-projects.github.io/micronaut-security/latest/guide/#ldap

### PR DESCRIPTION
**Pull request by polarising-java**
A first approach to resolve #101.
 Already implemented groups based on #143.

### CODE REVIEW

**Comment by polarising-java** [comment]
Add mock tests for Ldap: LdapAuthenticationProviderTest.java

File: docker-compose.yml [hunk 1]
```diff
@@ -105,8 +105,8 @@ services:
 
   kafkacat:
     image: confluentinc/cp-kafkacat
-    command: 
-      - bash 
+    command:
```
Comment (tchiotludo) [inline docker-compose.yml]: Please remove the whole file modification, not concerning the feature

File: src/main/java/org/kafkahq/modules/LdapContextAuthenticationMapper.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public AuthenticationResponse map(ConvertibleValues<Object> attributes, String username, Set<String> groups) {
+
+        List<String> kafkaHQgroups = getUserKafkaHQGroups(groups);
+        return new UserDetails(username, userGroupUtils.getUserRoles(kafkaHQgroups), userGroupUtils.getUserAttributes(kafkaHQgroups));
+    }
+
+    /**
+     * Get KafkaHQ Groups configured in Ldap groups
+     * @param ldapGroups
```
Comment (tchiotludo) [inline src/main/java/org/kafkahq/modules/LdapContextAuthenticationMapper.java]: please add comment description or remove `@param` & `@return`

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -30,11 +30,25 @@ micronaut:
     intercept-url-map:
       - pattern: "${kafkahq.server.base-path:}/static/**"
         access: "isAnonymous()"
+     # Ldap configuration
```
Comment (tchiotludo) [inline src/main/resources/application.yml]: Please remove the whole file modification, not concerning the feature

File: package-lock.json [hunk 4]
```diff
@@ -16,9 +16,9 @@
       "integrity": "sha512-1Bh06cbWJUHMC97acuD6UMG29nMt0Aqz1vF3guLfG+kHHJhy3AyohZFFxYk2f7Q1SQIrNwvncxAE0N/9s70F2w==",
       "dev": true,
       "requires": {
-        "@types/events": "*",
-        "@types/minimatch": "*",
-        "@types/node": "*"
+        "@types/events": "3.0.0",
```
Comment (tchiotludo) [inline package-lock.json]: Please remove the whole file modification, not concerning the feature

**Comment by tchiotludo** [comment]
Can you provide an example configuration that is working with https://www.forumsys.com/tutorials/integration-how-to/ldap/online-ldap-test-server/ for example ? 
I quickly try, and this is not working, must I failed the configuration.

Also, Can you add some configuration example on README.md (maybe with forumsys configurations) ?

File: src/main/resources/application.yml (hunk 3 above)
Comment (polarising-java) [reply src/main/resources/application.yml]: Removed changes

File: package-lock.json (hunk 4 above)
Comment (polarising-java) [reply package-lock.json]: Removed changes

File: docker-compose.yml (hunk 1 above)
Comment (polarising-java) [reply docker-compose.yml]: Rollback changes

File: src/main/java/org/kafkahq/modules/LdapContextAuthenticationMapper.java (hunk 2 above)
Comment (polarising-java) [reply src/main/java/org/kafkahq/modules/LdapContextAuthenticationMapper.java]: Added comments

**Comment by polarising-java** [comment]
I updated the README.md .
Tell me if you want more detail.

**Comment by tchiotludo** [comment]
I've changed `default-roles` with `default-groups` to remove globally roles.
I've made some changes on `README`
Also fixed a bug that won't displayed the login bar if you have only ldap authentification. 

But for now, I just can't make the ldap login work. 
As I quickly understand, the only provider that is used and tested is `BasicAuthAuthenticationProvider` . 
Look like the ldap provider is not used when I tried, maybe can you send me your full configuration in order to find what is wrong on my side ?

**Comment by polarising-java** [comment]
Try with my configuration:
[application.zip](https://github.com/tchiotludo/kafkahq/files/3883526/application.zip)

users (username/password - group):
riemann/password - mathematicians
einstein/password - scientists

First goes to BasicAuthAuthenticationProvider, and if no mtach then goes to LdapAuthenticationProdvider.

**Comment by tchiotludo** [comment]
ok !!! found it !! 
I'm a super Nooob !! :no_entry: 
Yaml space failed :) missing 2 one on ldap part ... 

Everything seems ok for me. 
If you just can confirm that my last modification (with groups) seems ok for you and you can merge it :fast_forward: :+1: :1st_place_medal:

**Comment by polarising-java** [comment]
Found also 2 situations:

- The default-groups doesn't apply the topics-filter-regexp

- if i remove default-groups from the configuration crashes with "Message: Error resolving field value [${kafkahq.security.default-groups}]. Property doesn't exist or cannot be converted"

**Comment by tchiotludo** [comment]
* default-groups / roles never allow `topics-filter-regexp` has I know ? I'm wrong ?  
* remove default-groups should not failed since I've added in the main configuration here 

https://github.com/polarising-public/kafkahq/blob/ce1a9d216e7bc06796e1b159adbbd102951e253b/src/main/resources/application.yml#L108

Can you send me the snippets that crash the app ?

**Comment by polarising-java** [comment]
If you remove default-groups from security section like this:

`security:
    groups:
      admin:
        roles:
        - topic/read
        - topic/insert
        - topic/delete
        - topic/config/update
        - node/read
        - node/config/update
        - topic/data/read
        - topic/data/insert
        - topic/data/delete
        - group/read
        - group/delete
        - group/offsets/update
        - registry/read
        - registry/insert
        - registry/update
        - registry/delete
        - registry/version/delete
        - acls/read
        - connect/read
        - connect/insert
        - connect/update
        - connect/delete
        - connect/state/update
      reader:
        roles:
          - topic/read
          - node/read
          - topic/data/read
          - group/read
          - registry/read
          - acls/read
          - connect/read`

**Comment by tchiotludo** [comment]
really strange, I can't reproduce on my side.
The default `application.yml` is used and It works for me. 
Do you have update with my last modification on this file ?

**Comment by polarising-java** [comment]
"default-groups / roles never allow topics-filter-regexp has I know ? I'm wrong ?"

You are right, but since we have default groups instead of default roles, and in a Group you can define a topics-filter-regexp, it's quite strange not applying the topic filter in the case of a default group.

**Comment by parmarques** [comment]
Added the topic filter regex also to default groups

**Pull request merged**